### PR TITLE
Implement a Software Pressure Control System for VTI Style Cryostats

### LINKED ISSUE #6947: Cryogenics: Implement the Enhanced Cryostat Control as IOC logic
As someone using cryogenics at ISIS (especially on EMU)  I need enhanced cryostat control scripts, so that systems running under pressure control will heat and cool automatically to the desired temperature, automatically adjusting the gas flow and heater for optimal running. 

### Acceptance Criteria
- [x] The python test script is enacted at the IOC level
- [x] ~~There is a desire to supply the definition of the pressure law as a Python function~~
    - Implemented as a linear interpolation using a text file similar to eurotherm ramp files. 
- [x] The following values should be user configurable:
   - [x] CUTOFF_POINT in Kelvin
   - [x] PRESSURE_SETPOINT_MIN in mbar
   - [x] PRESSURE_SETPOINT_MAX in mbar
   - [x] CONSTANT_PRESSURE_VALUE in mbar
   - [x] DELAY_BETWEEN_SETS in ms
   - [x] TEMP_SCALE
   - [x] PRESSURE_SET_BLOCK
   - [x] TEMP_BLOCK
   - [x] TEMP_SET_BLOCK
- [ ] PID values for the Mercury Pressure Control need to be sent to the device and managed in a similar fashion to the TPAR files
  - This is being implemented as a separate ticket.
- [x] Find a way to disambiguate this clearly from the LBC logic as they are both doing very similar things (controlling NV pressure based on temperature)
- [x] Ensure this logic and the "little blue cryostat" are mutually exclusive with each other

### Notes
- The design from #6729 is available at https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/MercuryiTC-Enhanced-Cryostat-Control and #6279 is likely to have other useful information
- The final script to be used to ensure the above is available in the EMU scrips as `cryo_control_test_v3.py`
- This is likely to need a state machine
- There is a test report available, which should be read in conjunction with the previous tickets to ensure that the requirements are understood well

## To Review
- [ ] https://github.com/ISISComputingGroup/EPICS-MercuryiTC/pull/22
- [ ] https://github.com/ISISComputingGroup/EPICS-ioc/pull/723
- [ ] https://github.com/ISISComputingGroup/EPICS-IOC_Test_Framework/pull/563
- [ ] https://github.com/ISISComputingGroup/ibex_gui/pull/1472
- [ ] https://github.com/ISISComputingGroup/EPICS-DeviceEmulator/pull/153
- [ ] https://github.com/ISISComputingGroup/EPICS-upgrade/pull/68
- [ ] https://github.com/ISISComputingGroup/IBEX/pull/7334

### PR DESCRIPTION
**Pull request by cailafinn**
## Work Done
- Added PVs to track the state of the statemachine and values that are relevant to the pressure control. Most of these are set by macros on IOC startup. 
- Added a statemachine to automatically change the pressure value based on the temperature and a number of other factors. The flowchart the statemachine is based on is [here](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/MercuryiTC-Enhanced-Cryostat-Control#flowchart)

Partially Resolves ISISComputingGroup/IBEX#6947

### CODE REVIEW

File: MercuryiTCApp/src/seqPVmacros.h [hunk 1]
```diff
@@ -0,0 +1,245 @@
+/*================================================================
```
Comment (rerpha) [inline MercuryiTCApp/src/seqPVmacros.h:1]: this file gets pulled in at compile time, so no need to store it in git

File: MercuryiTCApp/src/auto_pressure_ctrl.st [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+            if(temperature <= temperature_sp_rbv) {
+              new_pressure_sp = pressure_sp_min;
+            } else {
+              // Check cutoff for constant vs interpolation calc
+              if(temperature_sp_rbv <= temperature_cutoff && temperature <= temperature_cutoff) {
+                new_pressure_sp = pressure_const;
+              } else {
+                new_pressure_sp = pressure_sp_min + temperature_scale * pow((temperature - temperature_sp_rbv), 2);
+                new_pressure_sp = fmin(new_pressure_sp, pressure_max_lookup);
+              }
```
[the remark this answers is not in the document]
Comment (ConorMFinn) [reply MercuryiTCApp/src/auto_pressure_ctrl.st:102]: Ah, yes, that's intentional. There's an amendment in the minutes from a later meeting that explains why that I forgot to link on the ticket:
> A second issue related to the cooling of the system from high temperature (e.g. 100K) to base (less than ‘CUTOFF_POINT’). In this case line 46 of the code (which originally read ‘if temperature_setpoint <= cutoff_point’) resulted in the system immediately setting ‘constant_pressure_value’, irrespective of the cryostat temperature. If the cryostat temperature was well above the ‘CUTOFF_POINT’ when the set point was entered the system was unable to cool. Accordingly, this line was updated to ensure the ‘constant_pressure_value’ is only set once the cryostat temperature is below the ‘CUTOFF_POINT’.

File: MercuryiTCApp/src/auto_pressure_ctrl.st [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    state calc_new_sp {
+        entry
+        {
+            TRANSITION_TO_STATE("calc_new_sp");
+
+            if(temperature <= temperature_sp_rbv) {
+              new_pressure_sp = pressure_sp_min;
+            } else {
+              // Check cutoff for constant vs interpolation calc
+              if(temperature_sp_rbv <= temperature_cutoff && temperature <= temperature_cutoff) {
```
[the remark this answers is not in the document]
Comment (ConorMFinn) [reply MercuryiTCApp/src/auto_pressure_ctrl.st:97]: Not especially. We're not setting the temp setpoints as part of this logic, just reading them, so thought that in general it was suggested to use the RBV?

File: MercuryiTCApp/src/auto_pressure_ctrl.st (hunk 2 above)
[the remark this answers is not in the document]
Comment (ConorMFinn) [reply MercuryiTCApp/src/auto_pressure_ctrl.st:102]: I've added the changes made to the workflow (described by the testing document) to the linked wiki page

**Review by LilithCole — APPROVED** [review]

**Pull request merged**